File: tests/cpp/binary_trees/test_all_kinds_of_node_depths.cpp
#include "../../../problems/binary_trees/all_kinds_of_node_depths/cpp/all_kinds_of_node_depths.h"
#include "../jsontestbase.h"
#include "./utility.h"

class AllKindsOfNodeDepthsTest : public JsonTestBase {
public:
  AllKindsOfNodeDepthsTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/binary_trees/all_kinds_of_node_depths.json";
  }
};

TEST_F(AllKindsOfNodeDepthsTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded!";

  for (const auto &test : test_cases) {
    auto treeJson = test["tree"];
    auto expected = test["expected"].get<int>();

    auto root = buildBT(treeJson);

    auto result = allKindsOfNodeDepths(root);

    EXPECT_EQ(result, expected) << "Failed for test case: " << test.dump(2);

    deleteBT(root);
  }
}

File: tests/python/binary_trees/all_kinds_of_node_depths/test_all_kinds_of_node_depths.py
from problems.binary_trees.all_kinds_of_node_depths.python.all_kinds_of_node_depths import \
    allKindsOfNodeDepths
from problems.binary_trees.assets.BinaryTree import BinaryTree
from tests.python.binary_trees.utility import build_bt
from tests.python.utility import load_test_cases


def test_all_kinds_of_node_depths() -> None:

    test_cases = load_test_cases("binary_trees/all_kinds_of_node_depths.json")

    for idx, test in enumerate(test_cases):
        tree = test["tree"]
        expected = test["expected"]

        root: BinaryTree = build_bt(tree["nodes"], tree["root"])

        result = allKindsOfNodeDepths(root)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"
